sdk/schemaregistry/azure-data-schemaregistry-apacheavro/src/main/java/com/azure/data/schemaregistry/apacheavro/SchemaRegistrySchemaCache.java

```diff
@@ -0,0 +1,222 @@
+        }
+
+        final Mono<SchemaProperties> serviceCall;
+        if (autoRegisterSchemas) {
+            serviceCall = this.schemaRegistryClient
+                .registerSchema(schemaGroup, schemaFullName, schemaString, SchemaFormat.AVRO);
+        } else {
+            serviceCall = this.schemaRegistryClient.getSchemaProperties(
+                schemaGroup, schemaFullName, schemaString, SchemaFormat.AVRO);
+        }
+
+        return serviceCall.map(properties -> {
+            final String schemaId = properties.getId();
+
+            synchronized (lock) {
+                cache.put(schemaId, schema);
+                logCacheStatus();
+            }
+
+            return schemaId;
+        });
+    }
+
+    Mono<Schema> getSchema(String schemaId) {
+        synchronized (lock) {
+            final Schema existing = cache.get(schemaId);
+
+            if (existing != null) {
+                return Mono.just(existing);
+            }
+        }
+
+        return schemaRegistryClient.getSchema(schemaId)
+            .handle((registryObject, sink) -> {
+                final String schemaString = registryObject.getDefinition();
+
+                final Schema parsedSchema;
+                synchronized (lock) {
+                    parsedSchema = new Schema.Parser().parse(schemaString);
+                    cache.put(schemaId, parsedSchema);
+
+                    logCacheStatus();
+                }
+
+                sink.next(parsedSchema);
+            });
+    }
+
+    /**
+     * Gets number of cached schemas.
+     *
+     * @return Number of cached schemas.
+     */
+    int getSize() {
+        synchronized (lock) {
+            return cache.size();
+        }
+    }
+
+    /**
+     * Gets the length of schemas stored in cache.
+     *
+     * @return The length of schemas stored in cache.
+     */
+    int getTotalLength() {
+        synchronized (lock) {
+            return cache.getTotalLength();
+        }
+    }
+
+    /**
+     * Logs the cache status if log level verbose is enabled. Otherwise, no-op.
+     */
+    private void logCacheStatus() {
+        if (!logger.canLogAtLevel(LogLevel.VERBOSE)) {
+            return;
+        }
+
+        final int size = cache.size();
+        final int length = cache.getTotalLength();
+
+        logger.atVerbose()
+            .addKeyValue(SIZE_KEY, size)
+            .addKeyValue(TOTAL_LENGTH_KEY, length)
+            .log("Cache entry added or updated. Total number of entries: {}; Total schema length: {}",
+                size, length);
+    }
+
+    /**
+     * Simple LRU cache. Accesses to cache are synchronized via the outer class lock.
+     */
+    private static final class SchemaCache extends LinkedHashMap<String, Schema> {
+        private static final long serialVersionUID = -1L;
+
+        private final int capacity;
+        private final HashMap<Schema, String> schemaToIdCache = new HashMap<>();
```

Comment:
Why do we duplicate the entries in this HashMap and also have them in the LinkedHashMap?

Reply:
If we're trying to get the schemaId (for serialization) and we have the Schema, we'd have to do an O(n) search since the underlying map is keyed by the schemaId.